### ISSUE TITLE
Authentication of google service account with IMAP client failed

### ISSUE THREAD

**Issue by vaskezu** [jstedfast/MailKit#742]
Trying to authenticate a google service account using OAuth2. The service account have owner status with Google Drive and Gmail API enabled.

I'm following the instruction written on the FAQ page: https://github.com/jstedfast/MailKit/blob/master/FAQ.md#q-how-can-i-log-in-to-a-gmail-account-using-oauth-20

```csharp
await credential.RequestAccessTokenAsync(CancellationToken.None);
```
This method passes and I get the token is received successfully.

But the authentication fails with error MailKit.Security.AuthenticationException: 'Authentication failed.' on method:

```csharp
client.Authenticate(oauth2);
```

Got this in the IMAP client log:
```
Connected to imaps://imap.gmail.com:993/
S: * OK Gimap ready for requests from _**ip-address-is-here**_ q17-v6mb80211309wru
C: A00000000 CAPABILITY
S: * CAPABILITY IMAP4rev1 UNSELECT IDLE NAMESPACE QUOTA ID XLIST CHILDREN X-GM-EXT-1 XYZZY SASL-IR AUTH=XOAUTH2 AUTH=PLAIN AUTH=PLAIN-CLIENTTOKEN AUTH=OAUTHBEARER AUTH=XOAUTH
S: A00000000 OK Thats all she wrote! q17-v6mb80211309wru
C: A00000001 AUTHENTICATE XOAUTH2 _**a key is here**_
S: + 
_**a key is here**_
C: 
S: A00000001 NO Lookup failed q17-v6mb80211309wru
```

My whole code:

```csharp
        private async Task<ServiceAccountCredential> GetCredentialsWithAccessToken()
        {
            var certificate = new X509Certificate2(@".\my-secret.p12", "my-password", X509KeyStorageFlags.Exportable);
            var credential = new ServiceAccountCredential(new ServiceAccountCredential
                .Initializer("my-service-account.iam.gserviceaccount.com")
            {
                // Note: other scopes can be found here: https://developers.google.com/gmail/api/auth/scopes
                Scopes = new[] { "https://mail.google.com/" },
                User = "my-service-account.iam.gserviceaccount.com"
            }.FromCertificate(certificate));

            var success = await credential.RequestAccessTokenAsync(CancellationToken.None);
            return credential;
        }

        public async Task TestAuthentication()
        {
            using (var client = new ImapClient())
            {
                var credential = await GetCredentialsWithAccessToken();
                client.Connect("imap.gmail.com", 993, true);
                var oauth2 = new SaslMechanismOAuth2("my-service-account.iam.gserviceaccount.com", credential.Token.AccessToken);
                client.Authenticate(oauth2);
            }
        }
```

**Comment by jstedfast** [comment]
You need to use your gmail account name, not `"my-service-account.iam.gserviceaccount.com"`.

**Comment by vaskezu** [comment]
If I use the actual gmail account like in the code below, I get an exception on line: 
```csharp
var success = await credential.RequestAccessTokenAsync(CancellationToken.None);
```

Exception: 
```
Google.Apis.Auth.OAuth2.Responses.TokenResponseException: 'Error:"unauthorized_client", Description:"Client is unauthorized to retrieve access tokens using this method.", Uri:""'
```
```csharp
        private async Task<ServiceAccountCredential> GetCredentialsWithAccessToken()
        {
            var certificate = new X509Certificate2(@".\my-secret.p12", "my-password", X509KeyStorageFlags.Exportable);
            var credential = new ServiceAccountCredential(new ServiceAccountCredential
                .Initializer("my-service-account.iam.gserviceaccount.com")
            {
                // Note: other scopes can be found here: https://developers.google.com/gmail/api/auth/scopes
                Scopes = new[] { "https://mail.google.com/" },
                User = "my-account@gmail.com"
            }.FromCertificate(certificate));

            var success = await credential.RequestAccessTokenAsync(CancellationToken.None);
            return credential;
        }

        public async Task TestAuthentication()
        {
            using (var client = new ImapClient())
            {
                var credential = await GetCredentialsWithAccessToken();
                client.Connect("imap.gmail.com", 993, true);
                var oauth2 = new SaslMechanismOAuth2("my-account@gmail.com", credential.Token.AccessToken);
                client.Authenticate(oauth2);
            }
        }
```

Can I use this method of authentication if it's a normal ```@gmail.com``` account. Or this way is only for domain accounts?

**Comment by jstedfast** [comment]
This method works for normal gmail accounts.

The error you get when using account@gmail.com definitely means you are moving in the right direction. It's telling you that you don't have authorization for some reason.

I suspect that you might have registered for your auth token incorrectly or haven't registered for GMail access (e.g. maybe you just registered for Google Drive?).

Go back to https://console.developers.google.com/apis/dashboard and make sure you project includes GMail.

**Comment by spchaplin** [comment]
One critical step you must complete when trying to authenticate a Google service account using OAuth2, is that **_you must sign up for Google's paid service "G Suite"_**.  This enables you to delegate domain-wide authority to the service account to send emails attached to a domain you've verified with Google.  After going down the rabbit hole, I came to the conclusion that the complexity of setting this up with Google was not the best option for me.  I ended up using Mailgun, which provides the same service for free, and I found their documentation to be excellent.

Below are some links on the Google setup that may be helpful:

[https://stackoverflow.com/questions/42784640/client-is-unauthorized-to-retrieve-access-tokens-using-this-method-gmail-api-c-s/42785468](https://stackoverflow.com/questions/42784640/client-is-unauthorized-to-retrieve-access-tokens-using-this-method-gmail-api-c-s/42785468)

[https://developers.google.com/cloud-search/docs/guides/delegation](https://developers.google.com/cloud-search/docs/guides/delegation)
